File: PIA VPN/DataCounter.m
//
//  DataCounter.m
//  PIA VPN
//
//  Created by Davide De Rosa on 12/8/17.
//  Copyright © 2020 Private Internet Access Inc.
//
//  This file is part of the Private Internet Access iOS Client.
//
//  The Private Internet Access iOS Client is free software: you can redistribute it and/or
//  modify it under the terms of the GNU General Public License as published by the Free
//  Software Foundation, either version 3 of the License, or (at your option) any later version.
//
//  The Private Internet Access iOS Client is distributed in the hope that it will be useful,
//  but WITHOUT ANY WARRANTY; without even the implied warranty of MERCHANTABILITY
//  or FITNESS FOR A PARTICULAR PURPOSE.  See the GNU General Public License for more
//  details.
//
//  You should have received a copy of the GNU General Public License along with the Private
//  Internet Access iOS Client.  If not, see <https://www.gnu.org/licenses/>.
//

#include <arpa/inet.h>
#include <net/if.h>
#include <ifaddrs.h>
#include <net/if_dl.h>

#import "DataCounter.h"

BOOL DataCounterGetCurrentState(uint32_t *sent, uint32_t *received)
{
    struct ifaddrs *addrs;
    const struct ifaddrs *cursor;
    
    *sent = 0;
    *received = 0;
    
    if (getifaddrs(&addrs) != 0) {
        return NO;
    }
    
    cursor = addrs;
    while (cursor != NULL) {
        if (cursor->ifa_addr->sa_family == AF_LINK) {
            const struct if_data *ifa_data = (struct if_data *)cursor->ifa_data;
            if (ifa_data != NULL) {
//                DDLogVerbose(@"Data Counter: Interface name %s: sent %tu received %tu", cursor->ifa_name, ifa_data->ifi_obytes, ifa_data->ifi_ibytes);
            }
            
            // name of interfaces:
            // en0 is WiFi
            // pdp_ip0 is WWAN
            NSString *name = [NSString stringWithFormat:@"%s", cursor->ifa_name];
            if ([name hasPrefix:@"en"]) {
                const struct if_data *ifa_data = (struct if_data *)cursor->ifa_data;
                if (ifa_data != NULL) {
//                    WiFiSent += ifa_data->ifi_obytes;
//                    WiFiReceived += ifa_data->ifi_ibytes;
                    *sent += ifa_data->ifi_obytes;
                    *received += ifa_data->ifi_ibytes;
                }
            }
            
            if ([name hasPrefix:@"pdp_ip"]) {
                const struct if_data *ifa_data = (struct if_data *)cursor->ifa_data;
                if (ifa_data != NULL) {
//                    WWANSent += ifa_data->ifi_obytes;
//                    WWANReceived += ifa_data->ifi_ibytes;
                    *sent += ifa_data->ifi_obytes;
                    *received += ifa_data->ifi_ibytes;
                }
            }
        }
        
        cursor = cursor->ifa_next;
    }
    
    freeifaddrs(addrs);
    
    return YES;
}

File: PIA VPN/DataCounter.swift
//
//  DataCounter.swift
//  PIA VPN
//
//  Created by Davide De Rosa on 12/8/17.
//  Copyright © 2020 Private Internet Access Inc.
//
//  This file is part of the Private Internet Access iOS Client.
//
//  The Private Internet Access iOS Client is free software: you can redistribute it and/or
//  modify it under the terms of the GNU General Public License as published by the Free
//  Software Foundation, either version 3 of the License, or (at your option) any later version.
//
//  The Private Internet Access iOS Client is distributed in the hope that it will be useful,
//  but WITHOUT ANY WARRANTY; without even the implied warranty of MERCHANTABILITY
//  or FITNESS FOR A PARTICULAR PURPOSE.  See the GNU General Public License for more
//  details.
//
//  You should have received a copy of the GNU General Public License along with the Private
//  Internet Access iOS Client.  If not, see <https://www.gnu.org/licenses/>.
//

import Foundation
import PIALibrary

extension Notification.Name {
    static let DataCounterDidReport = Notification.Name("DataCounterDidReport")
}

class DataCounter {
    static let shared = DataCounter()

    private var isCounting = false

    private var timer: Timer?
    
    private var receivedBytes: UInt32 = 0
    
    private var sentBytes: UInt32 = 0

    private var lastStopDate: Date?

    private init() {
        let nc = NotificationCenter.default
        nc.addObserver(self, selector: #selector(applicationDidBecomeActive(notification:)), name: .UIApplicationDidBecomeActive, object: nil)
        nc.addObserver(self, selector: #selector(applicationWillResignActive(notification:)), name: .UIApplicationWillResignActive, object: nil)
    }
    
    func startCounting() {
        isCounting = true
        resumeCounting()
    }
    
    func resumeCounting() {
        guard (timer == nil) else {
            return
        }
        timer = Timer.scheduledTimer(
            timeInterval: AppConfiguration.DataCounter.interval,
            target: self,
            selector: #selector(timerDidCount),
            userInfo: nil,
            repeats: true
        )
    }
    
    func stopCounting() {
        timer?.invalidate()
        timer = nil
        lastStopDate = Date()
    }
    
    func notifyCurrentState() {
        timerDidCount()
    }
    
    // MARK: Notifications
    
    @objc private func timerDidCount() {
        guard Client.providers.accountProvider.isLoggedIn else {
            return
        }
        
        var numberOfSentBytes: UInt32 = 0
        var numberOfReceivedBytes: UInt32 = 0
        guard DataCounterGetCurrentState(&numberOfSentBytes, &numberOfReceivedBytes) else {
            return
        }
        
        let deltaReceived = UInt64(numberOfReceivedBytes - receivedBytes)
        let deltaSent = UInt64(numberOfSentBytes - sentBytes)
        
        let isFirst = ((receivedBytes == 0) && (sentBytes == 0))
        receivedBytes = numberOfReceivedBytes
        sentBytes = numberOfSentBytes
        
        // skip first calculation, no starting point for delta
        guard (!isFirst && (lastStopDate == nil)) else {
            lastStopDate = nil
            return
        }
        
        let downloadedBytes = deltaReceived
        let uploadedBytes = deltaSent
        
//        NSLog(@"DataCounter: D %.2fkb/s - U %.2fkb/s", downloadSpeed, uploadSpeed);
        Macros.postAppNotification(.DataCounterDidReport, [
            .downloaded: downloadedBytes,
            .uploaded: uploadedBytes
        ], false)
    }
    
    @objc private func applicationDidBecomeActive(notification: Notification) {
        if isCounting {
            resumeCounting()
        }
    }
    
    @objc private func applicationWillResignActive(notification: Notification) {
        stopCounting()
    }
}
